drink/src/lib.rs

```diff
@@ -77,7 +96,65 @@ impl<R: Runtime> Sandbox<R> {
     ///
     /// By default, a new `Sandbox` instance is created with a noop debug extension. This method
     /// allows to override it with a custom debug extension.
-    pub fn override_debug_handle(&mut self, d: DebugExt) {
+    pub fn override_debug_handle(&mut self, d: TracingExt) {
         self.externalities.register_extension(d);
     }
+
+    /// Registers the extension for intercepting calls to contracts.
+    fn setup_mock_extension(&mut self) {
+        self.externalities
+            .register_extension(InterceptingExt(Box::new(MockingExtension {
+                mock_registry: Arc::clone(&self.mock_registry),
+            })));
+    }
+}
+
+/// Runtime extension enabling contract call interception.
+struct MockingExtension<AccountId: Ord> {
+    /// Mock registry, shared with the sandbox.
+    ///
+    /// Potentially the runtime is executed in parallel and thus we need to wrap the registry in
+    /// `Arc<Mutex>` instead of `Rc<RefCell>`.
+    mock_registry: Arc<Mutex<MockRegistry<AccountId>>>,
+}
+
+impl<AccountId: Ord + Decode> InterceptingExtT for MockingExtension<AccountId> {
+    fn intercept_call(
+        &self,
+        contract_address: Vec<u8>,
+        _is_call: bool,
+        input_data: Vec<u8>,
+    ) -> Vec<u8> {
+        let contract_address = Decode::decode(&mut &contract_address[..])
+            .expect("Contract address should be decodable");
+
+        match self
+            .mock_registry
+            .lock()
+            .expect("Should be able to acquire registry")
+            .get(&contract_address)
+        {
+            // There is no mock registered for this address, so we return `None` to indicate that
+            // the call should be executed normally.
+            None => None::<()>.encode(),
+            // We intercept the call and return the result of the mock.
+            Some(mock) => {
+                let (selector, call_data) = input_data.split_at(4);
+                let selector: Selector = selector
+                    .try_into()
+                    .expect("Input data should contain at least selector bytes");
+
+                let result = mock
+                    .call(selector, call_data.to_vec())
+                    .expect("TODO: let the user define the fallback mechanism");
+
+                let result: ExecResult = Ok(ExecReturnValue {
+                    flags: ReturnFlags::empty(),
```

Comment:
Should we set the flags to `Revert` if the call failed?

Reply:
you mean: instead of the `expect` above?